### PR TITLE
Implement cloud function to merge prediction CSVs across chunks.

### PR DESCRIPTION
**Pull request by sylmak**

### CODE REVIEW

File: cloud_functions/climateiq_merge_output_cf/main.py
Comment (skeshive) [inline cloud_functions/climateiq_merge_output_cf/main.py:1]: rename this to climateiq_merge_scenario_predictions_cf?

File: cloud_functions/climateiq_merge_output_cf/main.py
Comment (skeshive) [inline cloud_functions/climateiq_merge_output_cf/main.py]: what do you mean by this?

File: cloud_functions/climateiq_merge_output_cf/main.py
Comment (skeshive) [inline cloud_functions/climateiq_merge_output_cf/main.py]: IMO we should use a separate bucket for export pipeline outputs. maybe "climateiq-spatialized-predictions"?

this bucket could contain the unmerged chunk outputs with object name: 
<prediction_type>/<model_id>/<study_area_name>/<scenario_id>/<chunk_id>

and the merged chunk outputs:
<prediction_type>/<model_id>/<study_area_name>/<chunk_id>

@waltaskew do you have any thoughts here?

File: cloud_functions/climateiq_merge_output_cf/main.py
Comment (skeshive) [inline cloud_functions/climateiq_merge_output_cf/main.py]: see comment below, we don't need to add the run_id prefix if we use a separate bucket

File: cloud_functions/climateiq_merge_output_cf/main.py
Comment (skeshive) [inline cloud_functions/climateiq_merge_output_cf/main.py]: setting this up to trigger via HTTP is fine for now but ideally we'd chain all the steps together to run automatically after the initial manual request to begin the export pipeline

the tricky part is figuring out when all the spatialization cloud functions have completed for every chunk

File: cloud_functions/climateiq_merge_output_cf/main.py
Comment (skeshive) [inline cloud_functions/climateiq_merge_output_cf/main.py]: could you add some function comments for all the functions

File: cloud_functions/climateiq_merge_output_cf/main.py
Comment (skeshive) [inline cloud_functions/climateiq_merge_output_cf/main.py]: in this case, I think we should actually throw and error and not output any merged CSVs for any chunks.

for all scenarios, we should have valid predictions for all chunks. if not, something went wrong so we shouldnt ship any new data

@waltaskew to confirm

File: cloud_functions/climateiq_merge_output_cf/main.py
Comment (waltaskew) [reply cloud_functions/climateiq_merge_output_cf/main.py]: Yes, I think a different bucket is a good idea. Otherwise, this cloud function would be triggered by its own writes to the same bucket!

File: cloud_functions/climateiq_merge_output_cf/main.py
Comment (waltaskew) [reply cloud_functions/climateiq_merge_output_cf/main.py]: I definitely agree we shouldn't ship data if some chunks are missing. We don't want to export data with holes in it! 

I think it's generally better to be brittle and fail loudly if data looks funny rather than try to recover and ship the funny looking data.

File: cloud_functions/climateiq_merge_output_cf/main.py
Comment (sylmak) [reply cloud_functions/climateiq_merge_output_cf/main.py]: The spatialized outputs should be in a different bucket (have made that change now) but should the merged chunk outputs be in a different bucket from the spatialized outputs? I can see a case for this bc it's easier to pick out the files.

File: cloud_functions/climateiq_merge_output_cf/main.py
Comment (sylmak) [reply cloud_functions/climateiq_merge_output_cf/main.py]: Changed to fail if a file isn't found. However, transactionally deleting all the files that have been written so far would be complex.

File: cloud_functions/climateiq_merge_output_cf/main.py
Comment (sylmak) [reply cloud_functions/climateiq_merge_output_cf/main.py]: Per Walt, the chunks for heat predictions will be per day (as opposed to flood, which are chunked by area). If we want to merge the heat and flood predictions together, it will take some additional transforms.

File: cloud_functions/climateiq_merge_output_cf/main.py
Comment (sylmak) [reply cloud_functions/climateiq_merge_output_cf/main.py]: This is a comment on your PR but my thinking is that we want to be able to trace the outputs to a specific batch ID in Vertex AI.

File: cloud_functions/climateiq_merge_output_cf/main.py
Comment (skeshive) [reply cloud_functions/climateiq_merge_output_cf/main.py]: we don't need to merge the heat and flood predictions together- just scenarios across both hazards. I confirmed this with Daniel/Jesus a while back

File: cloud_functions/climateiq_merge_output_cf/main.py
Comment (skeshive) [reply cloud_functions/climateiq_merge_output_cf/main.py]: I believe the model_id represents that based on an earlier conversation with Walt. A given version of a model would only run once per study_area/scenario_chunk. 

@waltaskew to confirm though

File: cloud_functions/climateiq_merge_output_cf/main.py
Comment (skeshive) [reply cloud_functions/climateiq_merge_output_cf/main.py]: Alternatively could we delete all the un-merged outputs and only store the merged outputs? I can't think of a use case where the un-merged outputs would be needed (aside from initial validation that the merging logic is working as intended)

File: cloud_functions/climateiq_merge_output_cf/main.py
Comment (skeshive) [reply cloud_functions/climateiq_merge_output_cf/main.py]: Could we add a clean up step before triggering the failure that deletes any file paths with no scenario id (indicating a merged output)?

Comment (waltaskew) [reply cloud_functions/climateiq_merge_output_cf/main.py]: I'd be less concerned about deleting files. As long as we know not to ship them, that's fine. The successful files could also be useful for debugging.
My hope would be that re-running once the issue is resolved would overwrite everything. Or folks could delete them manually. Either way, some manual intervention is going to be required to fix whatever bug caused the issue.

File: cloud_functions/climateiq_merge_output_cf/main.py
Comment (waltaskew) [reply cloud_functions/climateiq_merge_output_cf/main.py]: I think it's a question of if we re-ran the same model, would we want it to overwrite old outputs or produce new ones? Including the batch ID would do the latter.

File: cloud_functions/climateiq_merge_output_cf/main.py
Comment (sylmak) [reply cloud_functions/climateiq_merge_output_cf/main.py]: It's possible to set a TTL in GCS through the UI on a bucket level: http://cloud/storage/docs/lifecycle#age

We could put the merged and unmerged outputs in two different buckets, then set the unmerged output bucket to have a TTL of, say, 3 days.

Comment (sylmak) [reply cloud_functions/climateiq_merge_output_cf/main.py]: The downside would be that USL would have to make sure to run the merge function within that TTL.

We could also have this cloud function empty the unmerged bucket once merging is finished. Depends on where we want the complexity to live - in this cloud function, or in the GCS config.

File: cloud_functions/climateiq_merge_output_cf/main.py
Comment (sylmak) [reply cloud_functions/climateiq_merge_output_cf/main.py]: I think keeping the old outputs makes things easier to debug (especially since we haven't tested all this e2e yet). And there will never be a state where half the files in one set are from one run and half of them are from another (e.g., mid re-run). We don't need to retain the old outputs forever - we can set up a TTL and auto-delete the files once they reach a certain [age](https://cloud.google.com/storage/docs/lifecycle#age). The downside, as mentioned above, is that USL would have to ensure they trigger this cloud function before the files reach the TTL.

File: cloud_functions/climateiq_merge_output_cf/main.py
Comment (sylmak) [reply cloud_functions/climateiq_merge_output_cf/main.py]: Updated code to write to a different bucket.

File: cloud_functions/climateiq_merge_output_cf/main.py
Comment (skeshive) [reply cloud_functions/climateiq_merge_output_cf/main.py]: ^ lets remove this comment or reword it- its a bit confusing. rn all steps of the pipeline only handle flood predictions anyways

File: cloud_functions/climateiq_merge_scenario_predictions_cf/main.py
Comment (skeshive) [inline cloud_functions/climateiq_merge_scenario_predictions_cf/main.py]: climateiq-spatialized-chunk-predictions (see PR #22)

File: cloud_functions/climateiq_merge_scenario_predictions_cf/main.py
Comment (skeshive) [inline cloud_functions/climateiq_merge_scenario_predictions_cf/main.py]: climateiq-spatialized-merged-predictions

File: cloud_functions/climateiq_merge_scenario_predictions_cf/main.py
Comment (skeshive) [inline cloud_functions/climateiq_merge_scenario_predictions_cf/main.py]: this will just be a unique id...when we set up the output dir prefix for the vertex AI job, we don't have a job ID yet so the ML team will just be setting a unique id to group together outputs from the same job instead

File: cloud_functions/climateiq_merge_scenario_predictions_cf/main.py
Comment (skeshive) [inline cloud_functions/climateiq_merge_scenario_predictions_cf/main.py]: add TODO to look into automating this step instead of triggering via HTTP request

File: cloud_functions/climateiq_merge_scenario_predictions_cf/main.py
Comment (skeshive) [inline cloud_functions/climateiq_merge_scenario_predictions_cf/main.py:133]: is this necessary? wont the input blobs already match this pattern based on how we fetch them?

File: cloud_functions/climateiq_merge_scenario_predictions_cf/main.py
Comment (skeshive) [inline cloud_functions/climateiq_merge_scenario_predictions_cf/main.py:56]: lets preserve the file name ordering...id, prediction_type, model_id, study_area_name, chunk_id

File: cloud_functions/climateiq_merge_scenario_predictions_cf/main.py
Comment (skeshive) [inline cloud_functions/climateiq_merge_scenario_predictions_cf/main.py:61]: can you add a comment describing what these lines are doing here?

File: cloud_functions/climateiq_merge_scenario_predictions_cf/main.py
Comment (skeshive) [inline cloud_functions/climateiq_merge_scenario_predictions_cf/main.py]: with what headers?

File: cloud_functions/climateiq_merge_scenario_predictions_cf/main.py
Comment (skeshive) [inline cloud_functions/climateiq_merge_scenario_predictions_cf/main.py:38]: is this try...except needed? AFAIK, if a method raises an error and the caller doesn't handle it explicitly, the error would get raised?

File: cloud_functions/climateiq_merge_scenario_predictions_cf/main.py
Comment (skeshive) [inline cloud_functions/climateiq_merge_scenario_predictions_cf/main.py:27]: one more case to handle would be if all the H3 indexes are present for all scenarios. if any of the H3 cells are missing, we should throw an error

File: cloud_functions/climateiq_merge_scenario_predictions_cf/main.py
Comment (sylmak) [reply cloud_functions/climateiq_merge_scenario_predictions_cf/main.py:38]: Yes, this is to return the correct value for the HTTP response. I think if we just throw the error here the HTTP response will be a 500 instead.

File: cloud_functions/climateiq_merge_scenario_predictions_cf/main.py
Comment (sylmak) [reply cloud_functions/climateiq_merge_scenario_predictions_cf/main.py:133]: This is to account for files like batch/prediction_type/model_id/study_area_name/scenario_id/something_completely_unexpected/filename.unsupported_extension. We can also use the named match groups to extract the values.

**Review by skeshive — APPROVED** [review]

**Review by skeshive — COMMENTED** [review]
can also update the bucket names based on https://github.com/UrbanSystemsLab/climateiq-frontend/pull/25/files before updating?

**Pull request merged**
